### ISSUE TITLE
More flexibility in Naming

### ISSUE THREAD

**Issue by JacquesCarette** [JacquesCarette/Drasil#1901]
Indirectly in issue #1875 the problem of having a single global namespace arose. There, `p` is the usual notation for momentum, but it was already used.  We'd still like to be able to use `p` for momentum, at least when it doesn't clash.

At the very least, we ought to be able to give a *list* of notations for a particular concept. That still wouldn't help resolve the issue (it this case: if both position and momentum had alternate notations, how would one choose which to use the 'alternate' notation for?). Notation resolution should probably be a user-definable thing. What comes to mind is a concept-prioritization scheme, but that's just one idea.

What is really bad, and needs fixed, is the restriction to global uniqueness of displayed symbols. Per-document, and possibly per-system, uniqueness is needed. But not global!

**Comment by smiths** [comment]
I agree @JacquesCarette, especially with the statement that we should not restrict ourselves to global uniqueness of displayed symbols.  It is actually a bit amazing that we've made it this far with this constraint.  It is hard enough to not have symbol conflicts within one document.

**Comment by Mornix** [comment]
Could a “domain” of sorts be useful for this scenario?

For example, concepts (if I recall correctly) are tagged with the domain the knowledge resides in, could we further leverage this to prioritize symbol resolution?

As a simple example, take the imaginary number in Elec versus pure math. We could “assign” j for Elec domains and i for math domains.

Then in an SRS, an author could either specify or order their primary system domain or list of domains.

That would only leave symbols which are undefined in those domains and not available. The three thoughts I have for that is either an author “sets” them in the system configuration, we deterministically walk the domain tree (in some way, this is very train of thought at the moment) and find the “closest” domain with a free symbol, or Drasil performs some fresh variable renaming strategy with the closest matching symbol.

**Comment by smiths** [comment]
@Mornix, your idea is neat and clean.  I like it, but in its "pure" form, I don't think it will work.  Although there will be cases where a domain has a clear preference for a given symbol, this will not always be the case.  Not all concepts have "converged" to a standard symbol, even within a domain of knowledge.  There is still a problem in cross-domain work where the same symbol is used with different meanings.

I think we could use something like what you describe to "guess" what symbols would be appropriate, but then the user would need to have some say.  I'd like it if the user could override the symbol for their project.  We won't be able to predict all of the ways people will want to use symbols, so some kind of local override will be necessary.  I think your option of the author setting the symbols in the system configuration would work for this.

We currently cannot represent the same concept with different symbols, but do we check that we do not use the same symbol for different concepts?  I know we did not want to use p for two different concepts in #1875, but was that Drasil stopping us, or was it @oluowoj paying attention to what symbols had already been used?

**Comment by oluowoj** [comment]
Well, ideally, I thought it would be confusing to have one symbol represent different concepts in the same example, and I noticed this when I wanted to define momentum( I raised this concern as well last year with radius and displacement. 
I did not try defining multiple concepts with the same symbol, maybe Drasil will allow it, also not sure what the impact will be for code generation especially if these concepts are in the same domain.  

I agree with a user override option, that will allow a user to set the symbols.

**Comment by smiths** [comment]
Thank you for the clarification @oluowoj.  I would be surprised if Drasil detected that the same symbol is reused for different concepts.  It would be great if it did detect that, but you have confirmed my guess that it was you that noticed the symbol duplication.  We definitely do not want the same symbol reused for different concepts in the same document.  The same symbol for different concepts in different documents is fine, and really it is inevitable that we'll have to allow this.

**Comment by JacquesCarette** [comment]
Drasil won't detect that the same symbol is used twice in its database (and it shouldn't), but it ought to detect when the same symbol is used twice in the same document; well, it will certainly balk at the same UID being used twice, I'm less sure about the same symbol!  Feature request?

**Comment by smiths** [comment]
@JacquesCarette, yes I agree that we can reuse systems in our database.  We will have to!  There are not enough unique symbols to cover scientific and computing knowledge, especially if we aim for something understandable and match concepts and symbols according to their common use.  

I'll add an issue for the detection of the same symbol used for two different meanings in the same document.
